File: src/secsentry/reports/terminal.py
"""Plain-text report. One block per unique secret, not per line hit."""

from __future__ import annotations

import os
import sys

from secsentry.scan.engine import ScanReport

RESET = "\033[0m"
BOLD = "\033[1m"
DIM = "\033[2m"
COLOR = {
    "critical": "\033[91m",
    "high": "\033[93m",
    "medium": "\033[96m",
    "low": "\033[90m",
}


def _use_color() -> bool:
    if os.environ.get("NO_COLOR"):
        return False
    return sys.stdout.isatty()


def _paint(text: str, code: str, enabled: bool) -> str:
    return f"{code}{text}{RESET}" if enabled else text


def print_report(report: ScanReport) -> None:
    color = _use_color()
    secrets = report.secrets
    counts = report.counts()

    scanned = f"files={report.files_scanned}"
    if report.commits_scanned:
        scanned += f"  commits={report.commits_scanned}  blobs={report.blobs_scanned}"

    print(_paint("SecSentry", BOLD, color) + f"  {scanned}")
    summary = "  ".join(f"{n} {name}" for name, n in counts.items() if n)
    print(f"{len(secrets)} unique secret(s), {len(report.findings)} occurrence(s)" + (f"  —  {summary}" if summary else ""))
    if report.allowlisted:
        print(f"{report.allowlisted} finding(s) suppressed by .secsentryallow")
    print("-" * 70)

    if not secrets:
        print("No findings.")
        return

    for s in secrets:
        sev = s.severity.upper()
        head = _paint(f"[{sev}]", COLOR.get(s.severity.lower(), ""), color)
        print(f"{head} {_paint(s.secret_type, BOLD, color)}  {s.masked}  confidence={s.confidence:.2f}")

        for occ in s.occurrences[:10]:
            loc = f"{occ.path}:{occ.line}:{occ.column}"
            extra = f"  {occ.commit[:8]}" if occ.commit else ""
            print(f"    {loc}{extra}")
        if len(s.occurrences) > 10:
            print(f"    … {len(s.occurrences) - 10} more occurrence(s)")

        if s.introduced_by:
            when = s.first_seen[:10]
            age = f", {s.age_days}d ago" if s.age_days is not None else ""
            print(f"    introduced by {s.introduced_by} on {when}{age}")
        if s.still_in_head is not None:
            state = "still in HEAD" if s.still_in_head else "deleted from HEAD, still in history"
            print(f"    {state}")
        if s.why:
            print(_paint(f"    why: {'; '.join(s.why)}", DIM, color))
        print(_paint(f"    fix: {s.rotation}", DIM, color))
        print()

File: internal/report/terminal.go
package report

import (
	"fmt"
	"os"
	"strings"

	"github.com/umeraamir69/secsentry/internal/scan"
)

const (
	reset = "\033[0m"
	bold  = "\033[1m"
	dim   = "\033[2m"
)

var sevColor = map[string]string{
	"critical": "\033[91m",
	"high":     "\033[93m",
	"medium":   "\033[96m",
	"low":      "\033[90m",
}

func useColor() bool {
	if os.Getenv("NO_COLOR") != "" {
		return false
	}
	st, err := os.Stdout.Stat()
	if err != nil {
		return false
	}
	return st.Mode()&os.ModeCharDevice != 0
}

func paint(text, code string, on bool) string {
	if !on {
		return text
	}
	return code + text + reset
}

func Print(rep scan.Report) {
	color := useColor()
	secrets := rep.Secrets()
	counts := rep.Counts()

	scanned := fmt.Sprintf("files=%d", rep.FilesScanned)
	if rep.CommitsScanned > 0 {
		scanned += fmt.Sprintf("  commits=%d  blobs=%d", rep.CommitsScanned, rep.BlobsScanned)
	}
	fmt.Println(paint("SecSentry", bold, color) + "  " + scanned)

	var parts []string
	for _, name := range []string{"critical", "high", "medium", "low"} {
		if n := counts[name]; n > 0 {
			parts = append(parts, fmt.Sprintf("%d %s", n, name))
		}
	}
	line := fmt.Sprintf("%d unique secret(s), %d occurrence(s)", len(secrets), len(rep.Findings))
	if len(parts) > 0 {
		line += "  —  " + strings.Join(parts, "  ")
	}
	fmt.Println(line)
	if rep.Allowlisted > 0 {
		fmt.Printf("%d finding(s) suppressed by .secsentryallow\n", rep.Allowlisted)
	}
	fmt.Println(strings.Repeat("-", 70))
	if len(secrets) == 0 {
		fmt.Println("No findings.")
		return
	}
	for _, s := range secrets {
		head := paint("["+strings.ToUpper(s.Severity)+"]", sevColor[strings.ToLower(s.Severity)], color)
		fmt.Printf("%s %s  %s  confidence=%.2f\n", head, paint(s.SecretType, bold, color), s.Masked, s.Confidence)
		limit := len(s.Occurrences)
		if limit > 10 {
			limit = 10
		}
		for _, occ := range s.Occurrences[:limit] {
			loc := fmt.Sprintf("%s:%d:%d", occ.Path, occ.Line, occ.Column)
			extra := ""
			if occ.Commit != "" {
				c := occ.Commit
				if len(c) > 8 {
					c = c[:8]
				}
				extra = "  " + c
			}
			fmt.Printf("    %s%s\n", loc, extra)
		}
		if len(s.Occurrences) > 10 {
			fmt.Printf("    … %d more occurrence(s)\n", len(s.Occurrences)-10)
		}
		if s.IntroducedBy != "" {
			when := s.FirstSeen
			if len(when) > 10 {
				when = when[:10]
			}
			age := ""
			if s.AgeDays != nil {
				age = fmt.Sprintf(", %dd ago", *s.AgeDays)
			}
			fmt.Printf("    introduced by %s on %s%s\n", s.IntroducedBy, when, age)
		}
		if s.StillInHead != nil {
			if *s.StillInHead {
				fmt.Println("    still in HEAD")
			} else {
				fmt.Println("    deleted from HEAD, still in history")
			}
		}
		if len(s.Why) > 0 {
			fmt.Println(paint("    why: "+strings.Join(s.Why, "; "), dim, color))
		}
		fmt.Println(paint("    fix: "+s.Rotation, dim, color))
		fmt.Println()
	}
}
